main/src/ca/uwaterloo/flix/language/phase/Inliner1.scala

```diff
@@ -0,0 +1,986 @@
+        (MonoAst.Pattern.Tuple(ps, tpe, loc), subst)
+
+      case MonoAst.Pattern.Record(pats, pat, tpe, loc) =>
+        val (ps, substs) = pats.map(refreshRecordLabelPattern).unzip
+        val (p, subst) = refreshPattern(pat)
+        val subst1 = substs.foldLeft(subst)(_ ++ _)
+        (MonoAst.Pattern.Record(ps, p, tpe, loc), subst1)
+
+      case MonoAst.Pattern.RecordEmpty(tpe, loc) =>
+        (MonoAst.Pattern.RecordEmpty(tpe, loc), Map.empty)
+    }
+  }
+
+  /**
+    * Returns `true` if `def0` should be inlined.
+    */
+  private def canInlineDef(ctx: DefContext, context: Context): Boolean = {
+    val mayInline = ctx.occur != DontInline && !ctx.isSelfRecursive && context != Context.Stop
+    val shouldInline = ctx.isDirectCall ||
+      ctx.occur == Once ||
+      ctx.occur == OnceInAbstraction // May duplicate work?
+    mayInline && shouldInline
+  }
+
+  private def isPure(eff0: Type): Boolean = {
+    eff0.effects.isEmpty
+  }
+
+  /**
+    * Checks if `occur` is Dead.
+    */
+  private def isDead(occur: OccurrenceAst1.Occur): Boolean = occur match {
+    case Dead => true
+    case _ => false
+  }
+
+  /**
+    * Checks if `occur` is Once and `eff` is Pure
+    */
+  private def isUsedOnceAndPure(occur: OccurrenceAst1.Occur, eff0: Type): Boolean = occur match {
+    case Once => isPure(eff0)
+    case _ => false
+  }
+
+  /**
+    * Checks if `exp0` is trivial and `eff` is pure
+    */
+  private def isTrivialAndPure(exp0: MonoAst.Expr, eff0: Type): Boolean = {
+    isPure(eff0) && isTrivialExp(exp0)
+  }
+
+  /**
+    * Checks if `occur` is dead and `exp` is pure.
+    */
+  private def isDeadAndPure(occur: OccurrenceAst1.Occur, eff0: Type): Boolean = occur match {
+    case Dead => isPure(eff0)
+    case _ => false
+  }
+
+  /**
+    * Returns `true` if `exp0` is considered a trivial expression.
+    *
+    * An expression is trivial if:
+    * It is either a literal (float, string, int, bool, unit), or it is a variable.
+    *
+    * A pure and trivial expression can always be inlined even without duplicating work.
+    */
+  private def isTrivialExp(exp0: MonoAst.Expr): Boolean = exp0 match {
+    case MonoAst.Expr.Cst(_, _, _) => true
+    case MonoAst.Expr.Var(_, _, _) => true
+    case MonoAst.Expr.ApplyAtomic(AtomicOp.Unary(_), exps, _, _, _) => exps.forall(isTrivialExp)
+    case MonoAst.Expr.ApplyAtomic(AtomicOp.Binary(_), exps, _, _, _) => exps.forall(isTrivialExp)
+    case MonoAst.Expr.ApplyAtomic(AtomicOp.Tag(_), exps, _, _, _) => exps.forall(isTrivialExp)
+    case MonoAst.Expr.ApplyAtomic(AtomicOp.Tuple, exps, _, _, _) => exps.forall(isTrivialExp)
+    case _ => false
+  }
+
+  private def isHole(exp0: OccurrenceAst1.Expr): Boolean = exp0 match {
+    case OccurrenceAst1.Expr.ApplyAtomic(AtomicOp.Throw, _, _, _, _) => true
+    case OccurrenceAst1.Expr.ApplyAtomic(AtomicOp.HoleError(_), _, _, _, _) => true
+    case OccurrenceAst1.Expr.ApplyAtomic(AtomicOp.MatchError, _, _, _, _) => true
```

Comment:
hmmm match error is an interesting case - it makes every match expression cheat the effect system. They should be unreachable though. perhaps we should remove stop inserting the error case in the frontend of the compiler.